Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/PredictorCorrector.java

```diff
@@ -250,45 +266,53 @@ protected boolean trigger(DiVector candidate, int gap, int baseDimension, DiVect
-     * @param scale            the scale(s) of the multi-dimensional transformation;
-     *                         these would be mutiplied to any change in point or
-     *                         newPoint
-     * @param shift            the shift(s) of the multi-dimensional transoformation
-     * @param method           the transformation method (currently unusued, can be
-     *                         useful in future)
      * @param startPosition    the (most likely) location of the anomaly
-     * @param baseDimensions   the number of dimensions before shingling
+     * @param result           the result to be vended
      * @return true if the changes are significant (hence an anomaly) and false
      *         otherwise
      */
 
-    protected boolean isSignificant(boolean significantScore, double[] point, double[] newPoint, double[] scale,
-            double[] shift, TransformMethod method, int startPosition, int baseDimensions) {
+    protected <P extends AnomalyDescriptor> boolean isSignificant(boolean significantScore, double[] point,
+            double[] newPoint, int startPosition, P result) {
         checkArgument(point.length == newPoint.length, "incorrect invocation");
+        int baseDimensions = result.getDimension() / result.getShingleSize();
+        TransformMethod method = result.getTransformMethod();
+        boolean differenced = (method == TransformMethod.DIFFERENCE)
+                || (method == TransformMethod.NORMALIZE_DIFFERENCE);
+        double[] scale = result.getScale();
+        double[] shift = result.getShift();
+        // backward compatibility of older models
         if (scale == null || shift == null) {
             return true;
         }
         boolean answer = false;
-        for (int y = 0; y < baseDimensions && !answer; y++) {
-            double scaleFactor = (scale == null) ? 1.0 : scale[y];
-            double delta = Math.abs(point[startPosition + y] - newPoint[startPosition + y]) * scaleFactor;
-            double shiftBase = (shift == null) ? 0 : shift[y];
-            double shiftAmount = 0;
-            if (scaleFactor != 1.0 || shiftBase != 0) {
-                shiftAmount += 2 * DEFAULT_NORMALIZATION_PRECISION * Math.abs(shiftBase);
-            }
+        // only for input dimensions, for which scale is defined currently
+        for (int y = 0; y < result.getInputLength() && !answer; y++) {
+            double observedGap = Math.abs(point[startPosition + y] - newPoint[startPosition + y]);
+            double pathGap = calculatePathDeviation(point, startPosition, y, baseDimensions, differenced);
+            if (observedGap > 0.1 * pathGap) {
+                double scaleFactor = (scale == null) ? 1.0 : scale[y];
+                double delta = observedGap * scaleFactor;
+                double shiftBase = (shift == null) ? 0 : shift[y];
+                double shiftAmount = 0;
+                if (scaleFactor != 1.0 || shiftBase != 0) {
```

Comment:
It is, because the addition is multiplier * DEFAULT_NORMALIZATION_PRECISION * Math.abs(shiftBase) -- let me make that condition explicit (actually the condition is not even required because for Math.abs(shiftBase) == 0 the addition is 0. But this provides a place to tune the constants, if required.